prov/gni/src/gnix_mr_notifier.c

```diff
@@ -128,30 +130,38 @@ _gnix_notifier_close(struct gnix_mr_notifier *mrn)
 	int ret = FI_SUCCESS;
 	int ret_errno;
 
-	ret = notifier_verify_stuff(mrn);
+	fastlock_acquire(&mrn->lock);
 
-	if (ret == 0) {
-		fastlock_acquire(&mrn->lock);
+	ret = notifier_verify_stuff(mrn);
+	if (ret != FI_SUCCESS) {
+		GNIX_WARN(FI_LOG_MR, "Invalid MR notifier\n");
+		goto err_exit;
+	}
 
-		if (close(mrn->fd) != 0) {
-			ret_errno = errno;
-			GNIX_INFO(FI_LOG_MR, "error closing kdreg device: %s\n",
-				  strerror(ret_errno));
-			// Not all of these map to fi_errno values
-			ret = -ret_errno;
-			goto err_exit;
-		}
+	if (--mrn->ref_cnt) {
+		goto err_exit;
```

Comment:
I think this is best off with being a gnix reference object and controlled via the destruct. Given that the last reference to the notifier would be dropped when the last domain died, there isn't really a chance of a race with another open.